delegation/src/Pos/Delegation/Cede/Logic.hs

```diff
@@ -86,73 +89,61 @@ detectCycleOnAddition toAdd
                                  (isRevokePsk psk))
                    next
 
--- | Given a psk getPskr, issuer, delegate and cert he uses (to sign,
--- or taken from psk), checks if there's a psk chain "issuer →
--- delegate" and the last cert matches the provided one (can be
--- retrieved using 'psigPsk'). This *does not* check that delegate
--- didn't issue a psk to somebody else.
-dlgReachesIssuance
+-- | Returns the last certificate in the chain, starting with given
+-- issuer.
+--
+-- Delegate 'd' has right to issue block instead of issuer 'i' if
+-- there's a delegation chain:
+--
+-- i → x₁ → x₂ → … xₖ → d → ∅
+--
+-- where every arrow is resolved psk, and the last one xₖ → d
+-- equals to the passed one.
+dlgLastPsk
     :: (MonadCedeRead m)
-    => PublicKey                             -- ^ Issuer
-    -> PublicKey                             -- ^ Delegate
-    -> ProxySKHeavy                          -- ^ i->d psk
-    -> m Bool
-dlgReachesIssuance i d _ | i == d = pure True
-dlgReachesIssuance i d psk = reach i
+    => StakeholderId          -- ^ Issuer
+    -> m (Maybe ProxySKHeavy) -- ^ Last cert in the chain, if any
+dlgLastPsk i = reach (Nothing, i)
   where
-    -- Delegate 'd' has right to issue block instead of issuer 'i' if
-    -- there's a delegation chain:
-    --
-    -- i → x₁ → x₂ → … xₖ → d
-    --
-    -- where every arrow is resolved psk, and the last one xₖ → d
-    -- equals to the passed one.
-    reach curUser = getPsk (addressHash curUser) >>= \case
-        Nothing   -> pure False
-        Just psk'
-            | pskDelegatePk psk' == d -> pure $ psk' == psk
-            | otherwise               -> reach (pskDelegatePk psk')
+    reach (!prevPsk, !curUser) = getPsk curUser >>= \case
+        Nothing    -> pure prevPsk
+        Just !psk' -> reach (Just psk', addressHash $ pskDelegatePk psk')
 
 -- | Verifies a header from delegation perspective (signature checks).
 dlgVerifyHeader ::
        (MonadCedeRead m)
     => MainBlockHeader
     -> m (Either Text ())
 dlgVerifyHeader h = runExceptT $ do
-    -- Issuer didn't delegate the right to issue to elseone.
-    let issuer = h ^. mainHeaderLeaderKey
+    let bIssuer = h ^. mainHeaderLeaderKey
```

Comment:
Perhaps it'd be better to call this variable `leader` or `leaderPK`. Issuer is a bit ambiguous word: it can be block issuer or PSK issuer. Apparently `b` means block, but the leader doesn't necessary issue a block.
Moreover it's possible that the leader didn't issue anything. So calling her `bIssuer` is something I don't fully understand.

Reply:
Valid, fixed.